giza_actions/model.py

```diff
@@ -83,13 +81,14 @@ def __init__(
             self._get_credentials()
             self.model = self._get_model(id)
             self.version = self._get_version(version)
-            self.session = self._set_session()
             self.framework = self.version.framework
             self.uri = self._retrieve_uri()
             self.endpoint_id = self._get_endpoint_id()
+            if output_path:
+                self.session = self._set_session(output_path)
             if output_path:
```

Comment:
Having to pass `output_path` to `_download_model` and `_get_output_dtype` will mostly make `output_path` to be mandatory, but this is not the intention so another path where if `output_path` is not provided we handle it for the user:

* Lets make the class to have an attribute `self._output_path`
* If `output_path` is present when the user creates the instance, we will use that path, if not we will handle that
* Remove `output_path` argument from `_download_model` and use `self._output_path`
* Remove `output_path` from `_set_session` and `_get_output_dtype` as it is not needed any more

An example of the init could be:

```python
    def __init__(
        self,
        model_path: Optional[str] = None,
        id: Optional[int] = None,
        version: Optional[int] = None,
        output_path: Optional[str] = None,
    ):
       ...

        if model_path:
            self.session = ort.InferenceSession(model_path)
        elif id and version:
            self.model_id = id
            self.version_id = version
            self.model_client = ModelsClient(API_HOST)
            self.version_client = VersionsClient(API_HOST)
            self.api_client = ApiClient(API_HOST)
            self.endpoints_client = EndpointsClient(API_HOST)
            self._get_credentials()
            self.model = self._get_model(id)
            self.version = self._get_version(version)
            self.session = self._set_session()
            self.framework = self.version.framework
            self.uri = self._retrieve_uri()
            self.endpoint_id = self._get_endpoint_id()
            if output_path is not None:
                self._output_path = output_path
            else:
                self._output_path = os.path.join(tempfile.gettempdir(), f"{self.model_id}_{self.version_id}_{self.model.name})
            # Now this internally uses self._output_path
            # As we are using the cache hitting this function should not be problematic
            self._download_model()
```